src/cargo/sources/git/utils.rs

```diff
@@ -1007,13 +1016,36 @@ pub fn fetch(
     }
 
     if let Some(true) = gctx.net_config()?.git_fetch_with_cli {
-        return fetch_with_cli(repo, remote_url, &refspecs, tags, gctx);
+        match fetch_with_cli(repo, remote_url, &refspecs, tags, gctx) {
+            Ok(result) => return Ok(result),
+            Err(e) => {
+                return Err(e).context(format!("Revision {:?} not found", revision));
+            }
+        };
     }
 
     if gctx.cli_unstable().gitoxide.map_or(false, |git| git.fetch) {
-        return fetch_with_gitoxide(repo, remote_url, refspecs, tags, shallow, gctx);
+        match fetch_with_gitoxide(repo, remote_url, refspecs, tags, shallow, gctx) {
+            Ok(result) => return Ok(result),
+            Err(e) => {
+                if fast_path_rev == true {
+                    return Err(e).context(format!("Revision {:?} not found", revision));
```

Comment:
`.context` is eagerly evaluated — a string of `Revision ...` would be allocated even when it is not on the error path. use `with_context` to avoid that.